src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2299 @@
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        std = self.config.initializer_range
+        nn.init.normal_(self.alpha, mean=0.0, std=std)
+        nn.init.normal_(self.b_param, mean=0.0, std=std)
+
+    @staticmethod
+    def get_sinusoid_embeddings(max_positions: int, embedding_dim: int):
+        half_dim = embedding_dim // 2
+        emb = math.log(10000) / half_dim
+        emb = torch.exp(torch.arange(half_dim, dtype=torch.float) * -emb)
+        emb = torch.arange(max_positions, dtype=torch.float).unsqueeze(1) * emb.unsqueeze(0)
+        return torch.sin(emb), torch.cos(emb)
+
+    def rotary(self, x):
+        n, d = x.size()
+        x1, x2 = torch.chunk(x, 2, dim=-1)
+        if self.sine is None or n > self.sine.size(0):
+            self.sine, self.cosine = RotaryRelativePositionalBias.get_sinusoid_embeddings(n, d)
+            self.max_positions = n
+        self.sine = self.sine.to(self._float_tensor)
+        self.cosine = self.cosine.to(self._float_tensor)
+
+        sin = self.sine[:n]
+        cos = self.cosine[:n]
+        return torch.cat([x1 * cos - x2 * sin, x2 * cos + x1 * sin], dim=1)
+
+    def forward(self, seq_len):
+        a = self.rotary(self.alpha.expand(seq_len, self.embed_dim))
+        b = self.rotary(self.b_param.expand(seq_len, self.embed_dim))
+        t = torch.einsum("mk,nk->mn", a, b)
+        return t
+
+
+# Normalization modules
+# copied without modification
+class ScaleNorm(nn.Module):
+    def __init__(self, dim, eps=1e-6, affine=True):
+        super().__init__()
+        self.dim = dim
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.scalar = nn.Parameter(torch.Tensor(1))
+        else:
+            self.register_parameter("scalar", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.affine:
+            nn.init.constant_(self.scalar, 1.0)
+
+    def forward(self, x):
+        mean_square = torch.mean(torch.square(x), dim=self.dim, keepdim=True)
+        if self.scalar is not None:
+            x = self.scalar * x
+
+        x = x * torch.rsqrt(mean_square + self.eps)
+        return x
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, number_features, eps=1e-6, affine=True):
+        super().__init__()
+        self.num_features = number_features
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.weight = nn.Parameter(torch.Tensor(self.num_features))
+        else:
+            self.register_parameter("weight", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.affine:
+            nn.init.constant_(self.weight, 1.0)
+
+    def forward(self, x):
+        mean_square = torch.mean(torch.square(x), dim=-1, keepdim=True)
+        if self.weight is not None:
+            x = x * self.weight
+
+        x = x * torch.rsqrt(mean_square + self.eps)
+        return x
```

Comment:
are there required?

Reply:
yes, these are selected based on the option selected in [normalization_type](https://github.com/mnaylor5/transformers/blob/add-mega/src/transformers/models/mega/configuration_mega.py#L71-L73)